gpu4pyscf/qmmm/pbc/itrf.py

```diff
@@ -0,0 +1,1131 @@
+            ewpot0 += cp.einsum('ijxy,jxy->i', qm_ewald_hess[3], quads)
+            ewpot1  = cp.einsum('ijx,i->jx', qm_ewald_hess[1], charges)
+            ewpot1 += cp.einsum('ijxy,jy->ix', qm_ewald_hess[2], dips)
+            ewpot2  = cp.einsum('ijxy,j->ixy', qm_ewald_hess[3], charges)
+            return ewpot0, ewpot1, ewpot2
+
+        def get_hcore(self, mol=None):
+            cput0 = (logger.process_clock(), logger.perf_counter())
+            mm_mol = self.mm_mol
+            if mol is None:
+                mol = self.mol
+            rcut_hcore = mm_mol.rcut_hcore
+
+            if getattr(method_class, 'get_hcore', None):
+                h1e = cp.asarray(method_class.get_hcore(self, mol))
+            else:  # DO NOT modify post-HF objects to avoid the MM charges applied twice
+                raise RuntimeError('mm_charge function cannot be applied on post-HF methods')
+
+            Ls = mm_mol.get_lattice_Ls()
+            qm_center = np.mean(mol.atom_coords(), axis=0)
+
+            mask = np.linalg.norm(Ls, axis=-1) < 1e-12
+            Ls[mask] = [np.inf] * 3
+            r_qm = (mol.atom_coords() - qm_center)[None,:,:] - Ls[:,None,:]
+            r_qm = np.einsum('Lix,Lix->Li', r_qm, r_qm)
+            assert rcut_hcore**2 < np.min(r_qm), \
+                f"QM image is within rcut_hcore of QM center. " + \
+                f"rcut_hcore = {rcut_hcore} >= min(r_qm) = {np.sqrt(np.min(r_qm))}"
+            Ls[Ls == np.inf] = 0.0
+
+            r_qm = mol.atom_coords() - qm_center
+            r_qm = np.einsum('ix,ix->i', r_qm, r_qm)
+            assert rcut_hcore**2 > np.max(r_qm), \
+                f"Not all QM atoms are within rcut_hcore of QM center. " + \
+                f"rcut_hcore = {rcut_hcore} <= max(r_qm) = {np.sqrt(np.max(r_qm))}"
+            r_qm = None
+
+            qm_center = cp.asarray(qm_center)
+            all_coords = cp.asarray((mm_mol.atom_coords()[None,:,:] \
+                    + Ls[:,None,:]).reshape(-1,3))
+            all_charges = cp.hstack([mm_mol.atom_charges()] * len(Ls))
+            dist2 = all_coords - qm_center
+            dist2 = cp.einsum('ix,ix->i', dist2, dist2)
+
+            # charges within rcut_hcore exactly go into hcore
+            mask = dist2 <= rcut_hcore**2
+            charges = all_charges[mask]
+            coords = all_coords[mask]
+            logger.note(self, '%d MM charges see directly QM density'%charges.shape[0])
+            if mm_mol.charge_model == 'gaussian' and len(coords) != 0:
+                expnts = cp.hstack([mm_mol.get_zetas()] * len(Ls))[mask]
+                # FIXME slice mm coords when memory not enough
+                fakemol = gto.fakemol_for_charges(coords.get(), expnts.get())
+
+                intopt = int3c2e.VHFOpt(mol, fakemol, 'int2e')
+                intopt.build(self.direct_scf_tol, diag_block_with_triu=True, aosym=True, 
+                             group_size=int3c2e.BLKSIZE, group_size_aux=int3c2e.BLKSIZE)
+
```

Comment:
These calculation has been implemented in `get_j_int3c2e_pass1`. It is more efficient and consumes less memory.  And we may further improve this later.
https://github.com/pyscf/gpu4pyscf/blob/master/gpu4pyscf/solvent/pcm.py#L394

Reply:
herein I want to keep hcore w/o contracted with dm.

Reply:
You can contract the integrals with charges by `get_j_int3c2e_pass2`.